features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdh_alt.c

```diff
@@ -0,0 +1,256 @@
+
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_point_write_binary( grp, Q, MBEDTLS_ECP_PF_UNCOMPRESSED,
+                                                &public_key_size, temp_buf, sizeof(temp_buf) ) );
+
+        ret = convert_CrysError_to_mbedtls_err( CRYS_ECPKI_BuildPublKey( pDomain, temp_buf, public_key_size,
+                                                                         &ecdhParams->pubKey ) );
+        if ( ret != 0 )
+        {
+            goto cleanup;
+        }
+
+        memset ( temp_buf, 0, sizeof(temp_buf)  );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary( d, temp_buf, mbedtls_mpi_size( d ) ) );
+
+        ret = convert_CrysError_to_mbedtls_err( CRYS_ECPKI_BuildPrivKey( pDomain,
+                                                                         temp_buf,
+                                                                         mbedtls_mpi_size( d ),
+                                                                         &ecdhParams->privKey ) );
+        mbedtls_zeroize( temp_buf, sizeof( temp_buf ) );
+        if ( ret != 0 )
+        {
+            goto cleanup;
+        }
+
+        ret = convert_CrysError_to_mbedtls_err( CRYS_ECDH_SVDP_DH( &ecdhParams->pubKey, &ecdhParams->privKey,
+                                                                   secret, &secret_size,
+                                                                   &ecdhParams->ecdhTempData ) );
+        if ( ret != 0 )
+        {
+            goto cleanup;
+        }
+    }
+    else if ( grp->id ==  MBEDTLS_ECP_DP_CURVE25519 )
+    {
+        cc_ecc_25519_comp_shared_params_t* ecdhParams =  mbedtls_calloc( 1, sizeof(cc_ecc_25519_comp_shared_params_t) );
+        if ( ecdhParams == NULL )
+        {
+            ret = MBEDTLS_ERR_ECP_ALLOC_FAILED;
+            goto cleanup;
+        }
+
+        pHeap = ecdhParams;
+        heapSize = sizeof(cc_ecc_25519_comp_shared_params_t);
+
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary( d, ecdhParams->privKey, mbedtls_mpi_size( d ) ) ) ;
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary( &Q->X, ecdhParams->pubKey, public_key_size ) );
+
+        ret = convert_CrysError_to_mbedtls_err( CRYS_ECMONT_Scalarmult( secret, ( size_t* )&secret_size,
+                                                                        ecdhParams->privKey, CURVE_25519_KEY_SIZE ,
+                                                                        ecdhParams->pubKey, CURVE_25519_KEY_SIZE ,
+                                                                        &ecdhParams->kgTempData ) );
+        if ( ret != 0 )
+        {
+            goto cleanup;
+        }
+    }
+    else
+    {
+        ret =  MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
+        goto cleanup;
+    }
+
+    MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( z, secret, secret_size ) );
+
+cleanup:
+
+    if ( pHeap )
+    {
+        mbedtls_zeroize( pHeap, heapSize );
+        mbedtls_free ( pHeap );
+    }
+
+    if ( secret )
```

Comment:
I think this check is redundant because it would have returned [at the top](https://github.com/ARMmbed/mbed-os/pull/6794/files#diff-a4b7271ab3959058bdee1757841ea9afR154) if the allocation was not successful.

Reply:
right, but for readability I think it's better this way